crates/mcpclient/src/session.rs

```diff
@@ -0,0 +1,181 @@
+use crate::transport::{ReadStream, WriteStream};
+use crate::types::*;
+use anyhow::{anyhow, Context, Result};
+use serde::de::DeserializeOwned;
+use serde_json::{json, Value};
+use std::sync::atomic::{AtomicU64, Ordering};
+use tokio::sync::mpsc;
+
+struct OutgoingMessage {
+    message: JsonRpcMessage,
+    response_tx: mpsc::Sender<Result<Option<JsonRpcResponse>>>,
+}
+
+pub struct Session {
+    request_tx: mpsc::Sender<OutgoingMessage>,
+    id_counter: AtomicU64,
+}
+
+impl Session {
+    pub async fn new(read_stream: ReadStream, write_stream: WriteStream) -> Result<Self> {
+        let (request_tx, mut request_rx) = mpsc::channel::<OutgoingMessage>(32);
+
+        tokio::spawn(async move {
+            let mut pending_requests = Vec::new();
+            let mut read_stream = read_stream;
+            let write_stream = write_stream;
+
+            loop {
+                tokio::select! {
+                    Some(outgoing) = request_rx.recv() => {
+                        // Send the message
+                        if let Err(e) = write_stream.send(outgoing.message.clone()).await {
+                            let _ = outgoing.response_tx.send(Err(e.into())).await;
+                            continue;
+                        }
+
+                        // For requests, store the response channel for later
+                        if let JsonRpcMessage::Request(request) = outgoing.message {
+                            if let Some(id) = request.id {
+                                pending_requests.push((id, outgoing.response_tx));
+                            }
+                        } else {
+                            // For notifications, just confirm success
+                            let _ = outgoing.response_tx.send(Ok(None)).await;
+                        }
+                    }
+
+                    Some(message_result) = read_stream.recv() => {
+                        match message_result {
+                            Ok(JsonRpcMessage::Response(response)) => {
+                                if let Some(id) = response.id {
+                                    if let Some(pos) = pending_requests.iter().position(|(req_id, _)| *req_id == id) {
+                                        let (_, tx) = pending_requests.remove(pos);
+                                        let _ = tx.send(Ok(Some(response))).await;
+                                    }
+                                }
+                            }
+                            Ok(JsonRpcMessage::Notification(_)) => {
+                                // Handle incoming notifications if needed
+                            }
+                            Ok(_) => {
+                                println!("Unexpected message type");
+                            }
+                            Err(e) => {
+                                eprintln!("Error receiving message: {}", e);
```

Comment:
we probably shouldn't print all the errors here - need to propagate errors from the transport and then handle them here